runtime/runtime-params-estimator/src/replay/fold_db_ops.rs

```diff
@@ -0,0 +1,268 @@
+    /// Optionally collect and print detailed statistics for cache hits and misses.
+    track_caches: bool,
+    /// Keeps track of current block.
+    block_hash: Option<String>,
+    /// Stack of states, each starting at a specific indent.
+    states: Vec<State>,
+}
+
+/// Stores statistics for a range of the trace, starting at a fixed indent and
+/// collecting data for everything indented further, until another state object
+/// is pushed on top.
+#[derive(Default)]
+struct State {
+    /// The indent at which this state started
+    indent: usize,
+    /// Keeps track of operations per DB column.
+    ops_cols: BTreeMap<String, BTreeMap<String, usize>>,
+    /// Optionally collect and print detailed statistics for cache hits and misses.
+    cache_stats: Option<CacheStats>,
+}
+
+impl FoldDbOps {
+    pub(super) fn new() -> Self {
+        Self {
+            fold_anchors: HashMap::new(),
+            print_top_level: true,
+            account_filter: None,
+            track_caches: false,
+            states: vec![State::default()],
+            block_hash: None,
+            skip_indent: None,
+        }
+    }
+
+    /// Pre-set that folds on chunks.
+    pub(super) fn chunks(self) -> Self {
+        self.fold("apply_transactions", &["receiver", "shard_id"])
+    }
+
+    /// Pre-set that folds on receipts.
+    pub(super) fn receipts(self) -> Self {
+        self.fold("process_receipt", &["receiver", "receipt_id"])
+            .fold("process_transaction", &["receipt_id"])
+            .print_top_level(false)
+    }
+
+    pub(super) fn fold(mut self, anchor: impl Into<String>, printed_fields: &[&str]) -> Self {
+        let fields = printed_fields.into_iter().map(|s| (*s).into()).collect();
+        self.fold_anchors.insert(anchor.into(), fields);
+        self
+    }
+
+    pub(super) fn print_top_level(mut self, yes: bool) -> Self {
+        self.print_top_level = yes;
+        self
+    }
+
+    pub(super) fn with_cache_stats(mut self) -> Self {
+        self.track_caches = true;
+        self.state().cache_stats = Some(CacheStats::default());
+        self
+    }
+
+    pub(super) fn account_filter(mut self, account: Option<String>) -> Self {
+        self.account_filter = account;
+        self
+    }
+
+    fn state(&mut self) -> &mut State {
+        self.states.last_mut().expect(EMPTY_STATE_ERR)
+    }
+
+    fn push_state(&mut self, indent: usize) {
+        let cache_stats = if self.track_caches { Some(CacheStats::default()) } else { None };
+        let new_state = State { indent, ops_cols: Default::default(), cache_stats };
+        self.states.push(new_state);
+    }
+
+    fn pop_state(&mut self) -> State {
+        let state = self.states.pop().expect(EMPTY_STATE_ERR);
+        if self.states.is_empty() {
+            self.push_state(0);
+        }
+        state
+    }
+
+    fn skip(&mut self, indent: usize) -> bool {
+        if let Some(skip_indent) = self.skip_indent {
+            if skip_indent >= indent {
```

Comment:
Hm, shouldn't it be `if skip_indent > indent`? Perhaps it doesn't matter for our usecases, but I imagine there can be many ops with the same indent to skip, and the current code skips only one and don't skip others.

I would cover the expected behaviour in some test, which could be either synthetical or include receipts with receiver which must be skipped.